src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        currentState = Object.keys(currentState).reduce((newState, key) => {
```

Comment:
The use of `reduce` for removing properties is correct, but it may be less efficient than simply cloning the object and deleting the specified keys. Consider cloning `currentState` and then using a loop or `forEach` to delete the keys in `keysToRemove`.

Reply:
you can also use here 
`for (const key of keysToRemove) {
          if (stateCopy.hasOwnProperty(key)) {
            delete stateCopy[key];
          }
        }`